torch/distributed/_shard/checkpoint/utils.py

```diff
@@ -0,0 +1,193 @@
+from typing import Any, List, Union, Callable, Tuple, Optional
+import torch.distributed as dist
+from .api import CheckpointException
+import torch
+
+def tensor_narrow_n(tensor: torch.Tensor, offsets: Tuple[int, ...], lengths: Tuple[int, ...]) -> torch.Tensor:
+    for dim, (start, length) in enumerate(zip(offsets, lengths)):
+        tensor = torch.narrow(tensor, dim, start, length)
+    return tensor
+
+class _DistWrapper:
+    """
+    This is a wrapper around PG that provides a series of features around object collectives.
+
+    It works without distributed initialized, where most collectives turns into nops.
+
+    All variants that take functions are exception robust, meaning that if one or more
+    ranks raise errors, all ranks will observe those.
```

Comment:
I am wondering if we should build this robustness into the c10d library directly? seems pretty nice

Reply:
It's a huge lot harder to have that on the tensor collectives themselves.
This little wrapping only handles the case where the functional part fails.